### PR TITLE
Refactor part 2 - separate merge from diff iterator [change-log] optimize merge by 20 percent

### LINKED ISSUE #2784: Refactor the merge operation
Our code for the merge operation got to a level where it's too complex and it is very hard to follow when inserting changes. 

## Current Implementation 

As part of our design
- `Apply`  is used for both commit and merge
- Merge iterator is implemented using the Diff iterator 

The iterators used by `apply` for applying a merge are:
image

- Apply receives two iterators 
    1. The destination iterator 
    2. The compare iterator (only the changes we need to apply on destination)
 - Compare iterator wrapping the merge iterator in order to show only changes as records and ranges
 - Merge iterator 
 - Base Iterator used by the merge iterator in order to perform a three-way diff
 - Diff Iterator shows differences between two metarange iterators 
 - Left (Destination) iterator
 - Right ( source ) iterator 
 
Also, the merge iterator uses the terms `source` and `destination` and the diff iterator uses the terms `left` and `right`

## Suggestion 

We should try to separate between:
- Apply and Merge
- Diff and Merge

Examine if we could simplify by implementing merge using three Iterators (source destination and base) and The metarange writer. 

This suggestion might raise complexity elsewhere, we should first investigate and see if it could be simplified. 

This suggestion is should contain a detailed design

### PR DESCRIPTION
**Pull request by eden-ohana**
closes #2784
Done - Refactor part 1: separate commit and merge

Refactor part 2: 
* Separate merge ​from diff iterator (base, source, dest iterators instead)
*  Merge iterator is not used anymore by the merge (didn't delete because compare iterator)
*  Apply is deleted (both merge and commit doesn't use it anymore)
*  No longer supporting diff summary for merge 
* Allowing merge with no changes
* Optimize seek GE for base iterator

Next:
* Simplify the diff iterator by removing all extra functionality for merge
* Separate compare iterator from diff iterator

### CODE REVIEW

File: pkg/graveler/committed/merge.go [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+	switch {
+	case c < 0: // source before dest
+		baseValue, err := m.baseKeyGE(sourceValue.Key)
+		if err != nil {
+			return err
+		}
+		if baseValue != nil && bytes.Equal(sourceValue.Identity, baseValue.Identity) { // dest deleted this record
+			m.haveSource = m.source.Next()
+		} else {
+			if baseValue != nil && bytes.Equal(sourceValue.Key, baseValue.Key) { // conflict
```
Comment (guy-har) [inline pkg/graveler/committed/merge.go:228]:
```suggestion
			if baseValue != nil && bytes.Equal(sourceValue.Key, baseValue.Key) {  // deleted by dest and changed by source
```

File: pkg/graveler/committed/merge.go [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+		}
+	case c > 0: // dest before source
+		baseValue, err := m.baseKeyGE(destValue.Key)
+		if err != nil {
+			return err
+		}
+		if baseValue != nil && bytes.Equal(destValue.Identity, baseValue.Identity) { // source deleted this record
+			m.haveDest = m.dest.Next()
+		} else {
+			if baseValue != nil && bytes.Equal(destValue.Key, baseValue.Key) { // conflict
```
Comment (guy-har) [inline pkg/graveler/committed/merge.go:246]:
```suggestion
			if baseValue != nil && bytes.Equal(destValue.Key, baseValue.Key) { // deleted by source added by dest
```

File: pkg/graveler/committed/merge.go [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+					return fmt.Errorf("write record: %w", err)
+				}
+				m.haveSource = m.source.Next()
+				m.haveDest = m.dest.Next()
+				return nil
+			} else { // both changed the same key
+				return graveler.ErrConflictFound
+			}
+		}
+		// both added the same record
```
Comment (guy-har) [inline pkg/graveler/committed/merge.go:275]: Not precise, there is also the (more common) case of no change 
sourceValue.Identity == destValue.Identity == baseValue.Identity

File: pkg/graveler/committed/merge.go [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+	if bytes.Compare(destRange.MaxKey, sourceValue.Key) < 0 { // dest range before source
+		baseRange, err := m.baseRangeGE(graveler.Key(destRange.MinKey))
+		if err != nil {
+			return fmt.Errorf("base range GE: %w", err)
+		}
+		if baseRange != nil && destRange.ID == baseRange.ID { // source deleted this range
+			m.haveDest = m.dest.NextRange()
+			return nil
+		} else {
+			return graveler.ErrConflictFound // conflict
```
Comment (guy-har) [inline pkg/graveler/committed/merge.go:318]: No conflicts on ranges

File: pkg/graveler/committed/merge.go [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+		}
+	}
+	// source is at start of range which we need to scan, enter it
+	m.haveSource = m.source.Next()
+	return nil
+}
+
+func (m *merger) merge() error {
+	m.haveSource, m.haveDest, _ = m.source.Next(), m.dest.Next(), m.base.Next()
```
Comment (guy-har) [inline pkg/graveler/committed/merge.go:367]: why are you calling base.next without checking the result, what happens on error?

**Review by guy-har — CHANGES_REQUESTED** [review]

File: pkg/graveler/committed/merge.go [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+			}
+			if err != nil {
+				return err
+			}
+			m.haveSource = m.source.NextRange()
+			m.haveDest = m.dest.NextRange()
+		} else { // enter both ranges
+			m.haveSource = m.source.Next()
+			m.haveDest = m.dest.Next()
+		}
```
Comment (johnnyaug) [inline pkg/graveler/committed/merge.go:207]: IMO, more readable like this, but it's matter of preference:
```suggestion
		if baseRange == nil {
			fallthrough;
		}
		if sourceRange.ID == baseRange.ID {
			err = m.writeRange(destRange)
		} else if destRange.ID == baseRange.ID {
			err = m.writeRange(sourceRange)
		} else {
			fallthrough;
		}
		if err != nil {
			return err
		}
		m.haveSource = m.source.NextRange()
		m.haveDest = m.dest.NextRange()
```

File: pkg/graveler/committed/merge.go [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+		if baseRange == nil || destRange.ID == baseRange.ID { // source added this range
+			err = m.writeRange(sourceRange)
+			if err != nil {
+				return err
+			}
+			m.haveSource = m.source.NextRange()
+			return nil
+		}
+		return graveler.ErrConflictFound // conflict - both changed this range
```
[the remark this answers is not in the document]
Comment (johnnyaug) [reply pkg/graveler/committed/merge.go:167]: +add test case that covers this code

File: pkg/graveler/committed/merge.go [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+		baseRange, err := m.baseRangeGE(graveler.Key(sourceRange.MinKey))
+		if err != nil {
+			return fmt.Errorf("base range GE: %w", err)
+		}
+		if baseRange != nil && sourceRange.ID == baseRange.ID { // dest deleted this range
+			m.haveSource = m.source.NextRange()
+			return nil
+		}
+		if baseRange == nil || destRange.ID == baseRange.ID { // source added this range
+			err = m.writeRange(sourceRange)
```
Comment (johnnyaug) [inline pkg/graveler/committed/merge.go:160]: I think there's a bug here, adding a test case which I think will reflect it (add this to `Test_merge`).
I hope this is a valid case.
@guy-har, would love your input here as well.
```
		"is it a bug?": {
			baseRange: []testRange{
				{
					rng:     committed.Range{ID: "base:b-c", MinKey: committed.Key("b"), MaxKey: committed.Key("c"), Count: 2, EstimatedSize: 1024},
					records: []testValueRecord{{key: "b", identity: "b"}, {key: "c", identity: "c"}},
				},
				{
					rng:     committed.Range{ID: "base:d-e", MinKey: committed.Key("d"), MaxKey: committed.Key("e"), Count: 2, EstimatedSize: 1024},
					records: []testValueRecord{{key: "d", identity: "d"}, {key: "e", identity: "e"}},
				},
			},
			sourceRange: []testRange{
				{
					rng:     committed.Range{ID: "source:a-d", MinKey: committed.Key("a"), MaxKey: committed.Key("d"), Count: 2, EstimatedSize: 1024},
					records: []testValueRecord{{key: "a", identity: "a"}, {key: "b", identity: "b"}, {key: "d", identity: "d"}},
				},
				{
					rng:     committed.Range{ID: "source:e", MinKey: committed.Key("e"), MaxKey: committed.Key("e"), Count: 2, EstimatedSize: 1024},
					records: []testValueRecord{{key: "e", identity: "e1"}},
				},
			},
			destRange: []testRange{
				{
					rng:     committed.Range{ID: "base:b-c", MinKey: committed.Key("b"), MaxKey: committed.Key("c"), Count: 2, EstimatedSize: 1024},
					records: []testValueRecord{{key: "b", identity: "b"}, {key: "c", identity: "c"}},
				},
				{
					rng:     committed.Range{ID: "dest:d-e", MinKey: committed.Key("d"), MaxKey: committed.Key("e"), Count: 2, EstimatedSize: 1024},
					records: []testValueRecord{{key: "d", identity: "d1"}, {key: "e", identity: "e"}},
				},
			},
			conflictExpectedIdx: nil,
			expectedActions: []writeAction{
				{
					action:   actionTypeWriteRecord,
					key:      "a",
					identity: "source:a-d",
				},
				{
					action:   actionTypeWriteRecord,
					key:      "b",
					identity: "b",
				},
				{
					action:   actionTypeWriteRecord,
					key:      "d",
					identity: "d1",
				},
				{
					action:   actionTypeWriteRecord,
					key:      "e",
					identity: "e1",
				},
			},
		}
```

**Review by johnnyaug — CHANGES_REQUESTED** [review]
Impressive work!
I didn't finish reviewing yet, but since there is already one critical comment I'm publishing my review. Will continue tomorrow

File: pkg/graveler/committed/manager.go [hunk 9]
```diff
@@ -117,47 +105,42 @@ func (c *committedManager) Merge(ctx context.Context, ns graveler.StorageNamespa
...
-	summary := graveler.DiffSummary{
-		Count: map[graveler.DiffType]int{},
-	}
-	if err != nil {
-		return "", summary, fmt.Errorf("get metarange ns=%s id=%s: %w", ns, rangeID, err)
-	}
-	defer metaRangeIterator.Close()
-	summary, err = Apply(ctx, mwWriter, metaRangeIterator, diffs, &ApplyOptions{})
+
+	summary, err = Merge(ctx, mwWriter, baseIt, sourceIt, destIt)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/manager.go]: Done

File: pkg/graveler/committed/merge.go [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+
+	writer               MetaRangeWriter
+	base                 Iterator
+	source               Iterator
+	dest                 Iterator
+	haveSource, haveDest bool
+	summary              graveler.DiffSummary
+}
+
+// baseGE returns range from base iterator, which is greater or equal than the given key
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

File: pkg/graveler/committed/merge.go [hunk 11]
```diff
@@ -0,0 +1,425 @@
...
+		}
+	}
+	if err := m.base.Err(); err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+// baseGE returns value from base iterator, which is greater or equal than the given key
+func (m *merger) baseKeyGE(key graveler.Key) (*graveler.ValueRecord, error) {
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

File: pkg/graveler/committed/merge.go [hunk 12]
```diff
@@ -0,0 +1,425 @@
...
+			return baseRange, nil
+		}
+		if !m.base.NextRange() {
+			break
+		}
+	}
+	if err := m.base.Err(); err != nil {
+		return nil, err
+	}
+	return nil, nil
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

File: pkg/graveler/committed/merge.go [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+				}
+			}
+		} else if !m.base.NextRange() {
+			break
+		}
+	}
+	if err := m.base.Err(); err != nil {
+		return nil, err
+	}
+	return nil, nil
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

File: pkg/graveler/committed/merge.go [hunk 14]
```diff
@@ -0,0 +1,425 @@
...
+func (m *merger) baseKeyGE(key graveler.Key) (*graveler.ValueRecord, error) {
+	baseValue, _ := m.base.Value()
+	if baseValue != nil && bytes.Compare(key, baseValue.Key) <= 0 {
+		return baseValue, nil
+	}
+
+	for {
+		_, baseRange := m.base.Value()
+		if baseRange != nil && bytes.Compare(baseRange.MaxKey, key) >= 0 {
+			for {
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

File: pkg/graveler/committed/merge.go [hunk 15]
```diff
@@ -0,0 +1,425 @@
...
+			if !iter.Next() {
+				break
+			}
+		}
+	}
+	return iter.Err()
+}
+
+// applyBothRanges applies merging when both source and dest on range header
+func (m *merger) applyBothRanges(sourceRange *Range, destRange *Range) error {
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

File: pkg/graveler/committed/merge.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

File: pkg/graveler/committed/merge.go [hunk 16]
```diff
@@ -0,0 +1,425 @@
...
+				return err
+			}
+			m.haveDest = m.dest.NextRange()
+			return nil
+		}
+		if baseRange == nil || destRange.ID == baseRange.ID { // source deleted this range
+			m.haveDest = m.dest.NextRange()
+			return nil
+		}
+		return graveler.ErrConflictFound // conflict - both changed this range
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

File: pkg/graveler/committed/merge.go (hunk 1 above)
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

File: pkg/graveler/committed/merge.go (hunk 2 above)
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

File: pkg/graveler/committed/merge.go (hunk 3 above)
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

File: pkg/graveler/committed/merge.go (hunk 4 above)
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

File: pkg/graveler/committed/merge.go [hunk 17]
```diff
@@ -0,0 +1,425 @@
...
+		if iterValue == nil {
+			baseRange, err := m.baseRangeGE(graveler.Key(iterRange.MinKey))
+			if err != nil {
+				return fmt.Errorf("base range GE: %w", err)
+			}
+			if baseRange == nil || baseRange.ID != iterRange.ID {
+				if err := m.writeRange(iterRange); err != nil {
+					return err
+				}
+			}
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Changed this code

File: pkg/graveler/committed/merge.go (hunk 5 above)
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go:368]: Done

File: nessie/identity_test.go [hunk 18]
```diff
@@ -53,7 +53,7 @@ func TestIdentity(t *testing.T) {
 
 			resp, err := client.MergeIntoBranchWithResponse(ctx, repo, branch1, branch2, api.MergeIntoBranchJSONRequestBody{})
 			require.NoError(t, err, "error during merge")
-			require.NotNil(t, resp.JSON400, "merge should fail since there are no changes between the branches")
+			require.Nil(t, resp.JSON400, "allow merge with no changes between the branches")
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply nessie/identity_test.go]: Done

File: pkg/api/controller_test.go [hunk 19]
```diff
@@ -342,7 +342,10 @@ func TestController_CommitsGetBranchCommitLogByPath(t *testing.T) {
 		user:         "user3",
 		commitName:   "P",
 	})
-	mergeCommit, _ := deps.catalog.Merge(ctx, "repo3", "main", "branch-b", "user3", "commitR", nil)
+	mergeCommit, err := deps.catalog.Merge(ctx, "repo3", "main", "branch-b", "user3", "commitR", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/api/controller_test.go]: Done

File: pkg/graveler/committed/merge_test.go [hunk 20]
```diff
@@ -84,7 +86,6 @@ func Test_merge(t *testing.T) {
 		destRange           []testRange
 		expectedActions     []writeAction
 		conflictExpectedIdx *int
-		expectedSummary     graveler.DiffSummary
 		expectedErr         error
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge_test.go:89]: Done

File: pkg/graveler/committed/merge_test.go [hunk 21]
```diff
@@ -3,10 +3,12 @@ package committed_test
 import (
 	"bytes"
 	"context"
+	"errors"
 	"fmt"
 	"testing"
 
-	"github.com/go-test/deep"
+	"github.com/stretchr/testify/assert"
+
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge_test.go:11]: Done

File: pkg/graveler/committed/merge_test.go [hunk 22]
```diff
@@ -610,22 +610,224 @@ func Test_merge(t *testing.T) {
...
 				t.Fatal(err)
 			}
-			if diff := deep.Equal(summary, tst.expectedSummary); diff != nil {
-				t.Error("LoadActions() found diff", diff)
-			}
 		})
 	}
 }
+
+func TestMergeAdd(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge_test.go]: Done

File: pkg/graveler/committed/merge.go [hunk 23]
```diff
@@ -0,0 +1,425 @@
...
+		}
+	}
+	return iter.Err()
+}
+
+// applyBothRanges applies merging when both source and dest on range header
+func (m *merger) applyBothRanges(sourceRange *Range, destRange *Range) error {
+	switch {
+	case sourceRange.ID == destRange.ID: // source and dest added the same range
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Yes

File: pkg/graveler/committed/merge_test.go [hunk 24]
```diff
@@ -239,8 +235,14 @@ func Test_merge(t *testing.T) {
 			conflictExpectedIdx: nil,
 			expectedActions: []writeAction{
 				{
-					action: actionTypeWriteRange,
-					rng:    committed.Range{ID: "base:k1-k3", MinKey: committed.Key("k1"), MaxKey: committed.Key("k3"), Count: 2, EstimatedSize: 1024},
+					action:   actionTypeWriteRecord,
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge_test.go:238]: Yes, it's not a common case so I think we can skip it

File: pkg/graveler/committed/merge.go (hunk 8 above)
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go:206]: I fixed the bug and added this test, thanks!

File: pkg/graveler/committed/merge.go [hunk 25]
```diff
@@ -0,0 +1,425 @@
...
+	dest                 Iterator
+	haveSource, haveDest bool
+	summary              graveler.DiffSummary
+}
+
+// baseGE returns range from base iterator, which is greater or equal than the given key
+func (m *merger) baseRangeGE(key graveler.Key) (*Range, error) {
+	for {
+		_, baseRange := m.base.Value()
+		if baseRange != nil && bytes.Compare(baseRange.MaxKey, key) >= 0 {
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go:27]: Yes, It's a performance improvement. Instead of doing seekGE from the beginning of base iterator each time, it continues from its current location.

File: pkg/graveler/committed/merge_test.go [hunk 26]
```diff
@@ -586,8 +587,216 @@ func Test_merge(t *testing.T) {
...
+					records: []testValueRecord{{key: "b", identity: "b"}, {key: "c", identity: "c"}},
+				},
+				{
+					rng:     committed.Range{ID: "base:d-e", MinKey: committed.Key("d"), MaxKey: committed.Key("e"), Count: 2, EstimatedSize: 1024},
+					records: []testValueRecord{{key: "d", identity: "d"}, {key: "e", identity: "e"}},
+				},
+			},
+			sourceRange: []testRange{
+				{
+					rng:     committed.Range{ID: "source:a-d", MinKey: committed.Key("a"), MaxKey: committed.Key("d"), Count: 2, EstimatedSize: 1024},
```
Comment (guy-har) [inline pkg/graveler/committed/merge_test.go:693]: I'm confused, the test is called dest change key?

File: pkg/graveler/committed/merge_test.go [hunk 27]
```diff
@@ -586,8 +587,216 @@ func Test_merge(t *testing.T) {
 					rng:    committed.Range{ID: "base:k3-k6", MinKey: committed.Key("k3"), MaxKey: committed.Key("k6"), Count: 2, EstimatedSize: 1234},
 				},
 			},
-			expectedSummary: graveler.DiffSummary{Count: make(map[graveler.DiffType]int)},
-			expectedErr:     graveler.ErrNoChanges,
+			expectedErr: nil,
+		},
+		"source and dest change same range": {
```
Comment (guy-har) [inline pkg/graveler/committed/merge_test.go:592]:
```suggestion
		"dest removed range and added range after source removed range edges": {
```

Comment (guy-har) [reply pkg/graveler/committed/merge_test.go:592]: Not sure this is the best name, but it's a bit of a weird test, it's not that both changed the same range. It's that source changes are included in dest ranges.

File: pkg/graveler/committed/merge_test.go [hunk 28]
```diff
@@ -586,8 +587,216 @@ func Test_merge(t *testing.T) {
...
+					identity: "d1",
+				},
+				{
+					action:   actionTypeWriteRecord,
+					key:      "e",
+					identity: "e1",
+				},
+			},
+		},
+		"dest over source left": {
```
Comment (guy-har) [inline pkg/graveler/committed/merge_test.go:735]:
```suggestion
		"dest removed all source added": {
```

File: pkg/graveler/committed/merge_test.go [hunk 29]
```diff
@@ -586,8 +587,216 @@ func Test_merge(t *testing.T) {
...
+					action:   actionTypeWriteRecord,
+					key:      "a1",
+					identity: "a",
+				},
+				{
+					action:   actionTypeWriteRecord,
+					key:      "c",
+					identity: "c",
+				},
+			},
```
Comment (guy-har) [inline pkg/graveler/committed/merge_test.go:799]: Nice 👍

**Review by guy-har — CHANGES_REQUESTED** [review]
requesting changes because the fast forward optimization for `source == base` was removed

File: pkg/graveler/committed/manager.go [hunk 30]
```diff
@@ -92,74 +92,49 @@ func (c *committedManager) Diff(ctx context.Context, ns graveler.StorageNamespac
...
+	defer baseIt.Close()
 
-func (c *committedManager) Merge(ctx context.Context, ns graveler.StorageNamespace, destination, source, base graveler.MetaRangeID) (graveler.MetaRangeID, graveler.DiffSummary, error) {
-	summary := graveler.DiffSummary{
-		Count: map[graveler.DiffType]int{},
-	}
-	if source == base {
-		// no changes on source
-		return "", summary, graveler.ErrNoChanges
-	}
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/committed/manager.go]: You're right, it wasn't supposed to be removed. I added it back without error.

File: pkg/graveler/graveler.go [hunk 31]
```diff
@@ -640,7 +640,7 @@ type CommittedManager interface {
 	// Merge applies changes from 'source' to 'destination', relative to a merge base 'base' and
 	// returns the ID of the new metarange and a summary of diffs.  This is similar to a
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/graveler.go]: Done

File: pkg/graveler/committed/merge_test.go (hunk 26 above)
Comment (eden-ohana) [reply pkg/graveler/committed/merge_test.go:693]: changed its name

File: pkg/graveler/committed/merge_test.go (hunk 27 above)
Comment (eden-ohana) [reply pkg/graveler/committed/merge_test.go]: Done

File: pkg/graveler/committed/merge_test.go (hunk 28 above)
Comment (eden-ohana) [reply pkg/graveler/committed/merge_test.go]: Done

File: pkg/graveler/committed/merge.go [hunk 32]
```diff
@@ -0,0 +1,438 @@
...
+	case bytes.Compare(sourceRange.MaxKey, destRange.MinKey) < 0: // source before dest
+		baseRange, err := m.moveBaseToGERange(graveler.Key(sourceRange.MinKey))
+		if err != nil {
+			return fmt.Errorf("base range GE: %w", err)
+		}
+		if baseRange != nil && sourceRange.ID == baseRange.ID { // dest deleted this range
+			m.haveSource = m.source.NextRange()
+			return nil
+		}
+		if baseRange == nil || destRange.ID == baseRange.ID { // source added this range
```
Comment (johnnyaug) [inline pkg/graveler/committed/merge.go:208]: The check `destRange.ID == baseRange.ID` is irrelevant, and causes an optimization bug. At this point, `destRange` starts after `sourceRange` ends. Whether or not `baseRange` is the same range is not interesting. Only thing that matters is if baseRange intersects with sourceRange or not.

File: pkg/api/controller.go [hunk 33]
```diff
@@ -2791,37 +2791,13 @@ func (c *Controller) MergeIntoBranch(w http.ResponseWriter, r *http.Request, bod
...
-		case catalog.DifferenceTypeChanged:
-			result.Summary.Changed = v
-		case catalog.DifferenceTypeRemoved:
-			result.Summary.Removed = v
-		case catalog.DifferenceTypeConflict:
-			result.Summary.Conflict = v
-		}
-	}
-	return result
+	writeResponse(w, http.StatusOK, MergeResult{Reference: res}) // optimize returning unknown summary = 0
```
Comment (johnnyaug) [inline pkg/api/controller.go:2800]: Comment not clear - can be removed

File: pkg/graveler/committed/merge.go [hunk 34]
```diff
@@ -20,36 +20,45 @@ type merger struct {
...
-	}
-	baseRange, err := m.moveBaseToGERange(key)
-	if err != nil {
+	if err := m.base.Err(); err != nil {
 		return nil, err
 	}
+	return nil, nil
+}
+
+// getNextOverlappingFromBase moves base iterator (from current point) to the next overlap range with rangeToOverlap
```
Comment (johnnyaug) [inline pkg/graveler/committed/merge.go:51]:
```suggestion
// getNextOverlappingFromBase moves base iterator from its current position to the next range overlapping with rangeToOverlap
```

**Review by johnnyaug — APPROVED** [review]
Thanks for addressing all the comments. Looks very good!
Minor comments + would still love to see a benchmark justifying the existence of `getNextGEKey`

File: pkg/catalog/model.go [hunk 35]
```diff
@@ -46,7 +46,6 @@ type CommitLog struct {
 }
 
 type MergeResult struct {
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/catalog/model.go]: removed

File: pkg/api/controller.go (hunk 33 above)
Comment (eden-ohana) [reply pkg/api/controller.go]: Done

File: pkg/graveler/committed/merge.go (hunk 34 above)
Comment (eden-ohana) [reply pkg/graveler/committed/merge.go]: Done

**Comment by eden-ohana** [comment]
Thanks for the review  @guy-har @johnnyaug !
The refactor includes optimization in nextBaseGEkey - instead of moving the base iterator to the next key from the beginning each time, it moves from its current position (without scanning unnecessary ranges). 
Before the change the merge took 1.15m. After the optimization it took 56s.

**Comment by guy-har** [comment]
@eden-ohana 
## GREAT JOB 👍 
We really needed this refactor, Thanks!!
